src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_Wasm/UnboxingStubTargetMappingNode.cs

```diff
@@ -0,0 +1,142 @@
+    /// these nodes, which represent a tuple of { UnboxingStub, UnboxingStub.Target }), to be used by the runtime
+    /// to populate a hashtable that maps stubs to their targets. Note that we use this mechanism for both flavors
+    /// of unboxing stubs: ordinary and "special" (instantiating). The latter would ordinarily be stored alongside
+    /// the unwind info as "associated method data".
+    /// </summary>
+    public sealed class UnboxingStubTargetMappingNode : SortableDependencyNode
+    {
+        private readonly IMethodNode _unboxingStubNode;
+
+        public UnboxingStubTargetMappingNode(IMethodNode unboxingStubNode)
+        {
+            Debug.Assert(unboxingStubNode is UnboxingStubNode or ISpecialUnboxThunkNode { IsSpecialUnboxingThunk: true });
+            _unboxingStubNode = unboxingStubNode;
+        }
+
+        public IMethodNode Stub => _unboxingStubNode;
+        public ISymbolNode GetTarget(NodeFactory factor) => GetUnboxingStubTarget(factor, Stub);
+
+        protected override string GetName(NodeFactory factory) => $"Target mapping for {_unboxingStubNode.GetMangledName(factory.NameMangler)}";
+
+        protected override void OnMarked(NodeFactory factory) => factory.UnboxingStubTargetMappings.AddMapping(this);
+
+        public override bool StaticDependenciesAreComputed => true;
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context) => null;
+
+        public override bool InterestingForDynamicDependencyAnalysis => false;
+        public override bool HasDynamicDependencies => false;
+        public override IEnumerable<CombinedDependencyListEntry> SearchDynamicDependencies(List<DependencyNodeCore<NodeFactory>> markedNodes, int firstNode, NodeFactory context) => null;
+
+        public override bool HasConditionalStaticDependencies => false;
+        public override IEnumerable<CombinedDependencyListEntry> GetConditionalStaticDependencies(NodeFactory context) => null;
+
+        public override int ClassCode => 1371471006;
+
+        public override int CompareToImpl(ISortableNode other, CompilerComparer comparer)
+        {
+            // We rely on sorting regular unboxing stubs before instantiating ones.
+            IMethodNode stub = Stub;
+            IMethodNode otherStub = ((UnboxingStubTargetMappingNode)other).Stub;
+            if (stub.GetType() != otherStub.GetType())
+            {
+                return stub is UnboxingStubNode ? -1 : 1;
+            }
+
+            return comparer.Compare(stub, otherStub);
+        }
+
+        public static void AddConditionalMappingDependency(ref CombinedDependencyList dependencies, NodeFactory factory, IMethodNode unboxingStubNode)
+        {
+            // The mapping is only needed if the target will actually be present in the output.
+            if (factory.UseUnboxingStubTargetMappings && GetUnboxingStubTarget(factory, unboxingStubNode) is ISymbolNode target)
```

Comment:
I forget the flow of the node generation, are we guaranteed that the target dependency is added before this is called?

Reply:
No - that is why conditional dependencies are used. The mapping node will only be marked if both the stub and its target get marked.

(Or, if you meant "Is `GetUnboxingStubTarget` guaranteed to return a valid node always", then yes, the factory will create a new one if needed, even if it ends up not being marked)